### ISSUE TITLE
404 and infinite redirect on category pages in multistore mode

### ISSUE THREAD

**Issue by cewald** [DivanteLtd/vue-storefront#3396]
## Current behavior

If I got to a category page in a non-default store like `http://www.test.de/de/boys.html` the page will return a 404 (in multistore mode). But it will work in the default store without store-code in url like `http://www.test.de/de/boys.html`.

I first tracked it down to the changes for the new 404 page from #3296 & #3243, when the `beforeEachGuard` in [`core/modules/url/router/beforeEach.ts`](https://github.com/DivanteLtd/vue-storefront/blob/119be33841f1b5eec650d6e7764fc2c1fdf1c236/core/modules/url/router/beforeEach.ts#L30) is asking for `to.matched[0].name === 'page-not-found'` but the route name in multistore mode will be like `de-page-not-found`. So I changed it to `to.matched[0].name.endsWith('page-not-found')` to support multistores at this point.

Now it will serve the correct category page but will end in an infinite redirect in the browser (client-side) floating the memory.

The next thing I noticed is that the matched `dynamicRoutes` of `UrlDispatchMapper` are have url paths like `/uk/de/boys.html` after some infinite loops, so I looked at the `processDynamicRoute` method in [`core/modules/url/helpers/index.ts`](https://github.com/DivanteLtd/vue-storefront/blob/119be33841f1b5eec650d6e7764fc2c1fdf1c236/core/modules/url/helpers/index.ts#L18-L42) and fixed this by removing the store-code from path using `removeStoreCodeFromRoute` like:
```javascript
import { removeStoreCodeFromRoute } from '@vue-storefront/core/lib/multistore'
…

export function processDynamicRoute (routeData: LocalizedRoute, fullPath: string, addToRoutes: boolean = true): LocalizedRoute[] {
  fullPath = removeStoreCodeFromRoute(fullPath) as string
  ...
}
```
(I have mentioned this in https://github.com/DivanteLtd/vue-storefront/issues/3359#issuecomment-523325285 before – probably at the wrong place)

Now the url paths in the `dynamicRoutes` is correct, but there still is this infinite redirect on the client-side. It is always calling the `de-page-not-found` route, again and again. The server-side response and code is delivered correctly.

… I will try to track this down a little further and will keep you updated – but If you have any hints or already an idea what could cause this, please let me know.

## Expected behavior

Should also show category page with store-code in URL path.

## Steps to reproduce the issue

1. Set `storeViews.multistore` to `true`
1. Go to specific store category page like `http://localhost:3000/de/women/tops-women/jackets-women/jackets-23`
1. It should return a 404 response
1. If you go to the same page without a store-code like `http://localhost:3000/women/tops-women/jackets-women/jackets-23`, it should work

## Repository

develop (since 3d2b4a2e69b8e0707f7ecd58c320fb0172a27271)

## Can you handle fixing this bug by yourself?

- [ ] YES (I try, and will let you know if I could)
- [x] NO

## Which [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html) state this refers to? Info for developer.
Pick one option.

- [x] This is a bug report for test version on https://test.storefrontcloud.io - In this case Developer should create branch from `develop` branch and create Pull Request `2. Feature / Improvement` back to `develop`.
- [ ] This is a bug report for current Release Candidate version on https://next.storefrontcloud.io - In this case Developer should create branch from `release` branch and create Pull Request `3. Stabilisation fix` back to `release`.
- [ ] This is a bug report for current Stable version on https://demo.storefrontcloud.io and should be placed in next stable version hotfix - In this case Developer should create branch from `hotfix` or `master` branch and create Pull Request `4. Hotfix` back to `hotfix`.

## Environment details

- Browser: Chrome
- OS: OSX
- Node: 10 LTS
- Code Version: 1.11.0-rc.1 - develop 3d2b4a2e69b8e0707f7ecd58c320fb0172a27271

**Comment by cewald** [comment]
I did some more research, and if I see it right, it always matches the `*` route of `page-not-found` for dynamic routes (`routeData`) results in `beforeEachGuard`.

For example:  if there is a category found it will return the correct `routeData` like `{ name: 'category', params: … }` but also returns just one item in its `matches` array which is `page-not-found` because of the `/*` path. Does this make sense?

I could not find a solution yet, so, any help would be very appreciated :)

**Comment by pkarw** [comment]
@cewald it might be as you wrote - vue.js executes the guards chain by each `addRoutes()` call, moreover there could be a multimatch (first the right route then `*` - maybe there is a way to prevent it) @lukeromanowicz it's been seemingly introduced by the #3243

**Comment by ResuBaka** [comment]
Then I can maybe give some infos on where the error could come from. 

I have found today that in [here](https://github.com/DivanteLtd/vue-storefront/blob/b6d713725030b0b9c6e489cc447a8e7ba28368af/core/server-entry.ts#L66) the  `router.onReady(() => {` is called now later than in 1.9 I have not tested it in 1.10.
So that router guards are called before `router.onReady`  this could be the problem in the end.

So could you maybe test [this code change](https://www.diffchecker.com/foIJxvjY)  maybe that help. 
When it helps we need to find the root cause of this.

**Comment by cewald** [comment]
@ResuBaka I tested your code on `1.11.0-rc.1` and it didn't resolve the problem. As far as I understand the storeView is prepared correctly when the `mappingFallback` and `mapUrl` is called. The results of it are right, except of the matched route, which is always `page-not-found` with `*`, which is actually correct. But shouldn't it also match the fallback routes?

**Comment by ResuBaka** [comment]
So you have a fallback URL in your routes? 

One other problem could be that now all router guards from modules are fired before the theme routes are registered could that be the problem you have? @cewald

**Comment by cewald** [comment]
@ResuBaka Yes a fallback is been returned. E.g.: If I went to `http://localhost:3000/de/women/tops-women/jackets-women/jackets-23`, I get the following `routeData` from `mapUrl`/`mappingFallback` dispatch:
```json5
{
  fullPath: "/de/women/tops-women/jackets-women/jackets-23",
  hash: "",
  matched: [
    0: {path: "*", regex: /^((?:.*))(?:\/(?=$))?$/i, components: {…}, instances: {…}, name: "page-not-found", …}
  ],
  meta: {},
  name: "de-category",
  params: {slug: "jackets-23"},
  path: "/de/women/tops-women/jackets-women/jackets-23",
  query: {}
}
```
I'm currently testing in a plain VSF installation of latest develop branch where I enabled the multistore mode.

**Comment by cewald** [comment]
@pkarw @ lukeromanowicz What could be good solution?

Would it be convenient to create a state value like `lastRouteMatch` to evaluate if the current `beforeEachGuard.to` value isn't the same as the last to prevent this infinite match loop?

Or maybe it's somehow possible to rearrange the routes that the `*`/`page-not-found` route is always the last one as suggested normally.

**Comment by cewald** [comment]
I've tested the first approach where I added an extra state value which is set in `beforeEachGuard` and it seems to work. But I'm not really sure if this is a good practice.  
See it in detail here: https://github.com/DivanteLtd/vue-storefront/compare/DivanteLtd:develop...icmaa:bugfix/3396-wrong-404-redirect

**Comment by lukeromanowicz** [comment]
@cewald your fix seems to be a fine workaround but it definitely should not be considered as the final solution. Instead of finding and eliminating the root cause, we're only limiting the symptoms. Did you have any luck with finding why de-page-not-found is being dispatched over and over?

**Comment by cewald** [comment]
@lukeromanowicz Yes, the second part is definitely just a workaround.

I guess the problem is that the `page-not-found` route is always matched before the custom dynamic routes, because the custom routes are added always to the end of the routes array via [`router.addRoutes`](https://router.vuejs.org/api/#router-addroutes). But I didn't found a way to change this order. Nevertheless, it is working this way on single store mode.

Another idea is probably to change the route path back to `/page-not-found` and redirect to it like before, except using `next({'name': 'page-not-found', replace: 'false'})` to not change the url-path.

**Comment by pkarw** [comment]
Unfortunately there is probably no way to re-order the routes other than messing with Vue.router internals which is probably not something we’d like to do :)

Maybe there is a way to just move the place where ‘*’ is being added to after modules and themes registration (in app.ts) - as modules and themes are adding the routes in their registration sequence

Not really sure about this as the urlDispatcher is also adding the routes dynamically - so you can never be sure if the ‘*’ is really at the very end

Another idea is to add exactly the same solution for 404 as we have for normal dynamic routes: add another dynamic route to page 404 - and that’s all!

I mean - to remove the ‘*’ route and refactor https://github.com/DivanteLtd/vue-storefront/blob/54583355aabd3334c048124ea9a625ae1f7203ce/core/modules/url/router/beforeEach.ts#L37

- instead of doing redirect to page-not-found call the `processDynamicRoute` to add a route to page 404 fo this specific URL address; the `page-not-found` route should be passed as `routeData` parameter

**Comment by cewald** [comment]
@pkarw I tried your approach like: https://github.com/DivanteLtd/vue-storefront/compare/DivanteLtd:develop...icmaa:bugfix/3396-wrong-404-redirect
I added a dynamic 404-dispatcher for the current url if the path is not mappable. It weirdly kind of leads to a similar problem. The page will be now loaded correctly if found, but if not it shows the 404-page for a second and then disappears, leaving behind an empty content section. One step closer :)

But @lukeromanowicz seems to be close to solution.

**Comment by ResuBaka** [comment]
The PR that @lukeromanowicz has opened works for the and him. So it looks like it is fixed, I am only going to check the PR if it is all still working when you use host based storeViews and not storeCode based ones.

Could you test if that branch fixes it for you too? @cewald

**Comment by cewald** [comment]
I tried it. It seems to work – but if I go to a page which leads to a 404-page, the 404-page shortly appears in the browser and then disappears again and left behind an empty content section – or, on a plain installation, the 404-page also disappears for a second and is back again after half a second (like flickering). @ResuBaka Can you relate to this flickering?

**Comment by lukeromanowicz** [comment]
I believe that my fix is working correctly on 201e470896e91cd348f21d8cfc2474bf128d2d62. Later the latest develop was merged and caused some new issues. I'm tracking down the root of it but its not directly related. I have a confirmation from @cewald and @alinadivante that before merging latest develop it was working fine.

**Comment by pkarw** [comment]
Great job! I really liked the teamwork spirit shown by this issue. So, cool! :)
